elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/metadata/SQLReferenceTable.java

```diff
@@ -134,12 +157,39 @@ public void resolveAndStoreAllReferencesAndJoins(Queryable queryable) {
             resolvedJoinProjections.put(key, new HashMap<>());
         }
 
-        FormulaValidator validator = new FormulaValidator(metaDataStore);
+        if (!globalTablesContext.containsKey(key)) {
+
+            if (!isNested) {
+                Type<?> entityClass = dictionary.getEntityClass(key.getName(), key.getVersion());
+                SQLTable table = (SQLTable) metaDataStore.getTable(entityClass);
+
+                TableContext tableCtx = new TableContext(key.getAlias(),
+                                                         dialect,
+                                                         prepareColumnArgsMap(table.getArguments(), TBL_PREFIX));
+                globalTablesContext.put(key, tableCtx);
+                Set<Type<?>> visitedTableContext = new HashSet<>();
+                populateTableContext(entityClass, tableCtx, StringUtils.EMPTY, visitedTableContext);
+            } else {
+                TableContext tableCtx = new TableContext(key.getAlias(),
+                                                         dialect,
+                                                         Collections.emptyMap());
+                globalTablesContext.put(key, tableCtx);
+                queryable.getColumnProjections().forEach(column -> {
+                    tableCtx.put(column.getName(), new ColumnContext(column.getExpression(),
+                                                                     StringUtils.EMPTY,
+                                                                     Collections.emptyMap()));
+                });
+            }
+        }
+
         SQLJoinVisitor joinVisitor = new SQLJoinVisitor(metaDataStore);
 
         queryable.getColumnProjections().forEach(column -> {
-            // validate that there is no reference loop
-            validator.visitColumn(queryable, column);
+            if (!isNested) {
+                // validate that there is no reference loop
+                FormulaValidator validator = new FormulaValidator(globalTablesContext.get(key));
```

Comment:
With optimization, joins can occur in the outer or inner queries - hence a loop can appear in an outer query.

Reply:
Added validation for nested queries.